src/VisualStudio/Core/Def/Implementation/CodeLens/VisualStudioCodeLensReferencesService.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeLens;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.CodeLens
+{
+    [ExportWorkspaceService(typeof(ICodeLensReferencesService), layer: ServiceLayer.Host), Shared]
+    internal sealed class VisualStudioCodeLensReferencesService : ICodeLensReferencesService
+    {
+        public async Task<ReferenceCount> GetReferenceCountAsync(Solution solution, DocumentId documentId, SyntaxNode syntaxNode, int maxSearchResults,
+            CancellationToken cancellationToken)
+        {
+            var remoteService = solution.Workspace.Services.GetService<IRemoteCodeLensReferencesService>();
+            return remoteService == null
+                ? await
+                    CodeLensReferencesServiceFactory.Instance.GetReferenceCountAsync(solution, documentId, syntaxNode,
+                        maxSearchResults, cancellationToken).ConfigureAwait(false)
+                : await
+                    remoteService.GetReferenceCountAsync(solution, documentId, syntaxNode, maxSearchResults,
+                        cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<IEnumerable<ReferenceLocationDescriptor>> FindReferenceLocationsAsync(Solution solution, DocumentId documentId, SyntaxNode syntaxNode,
+            CancellationToken cancellationToken)
+        {
+            var remoteService = solution.Workspace.Services.GetService<IRemoteCodeLensReferencesService>();
```

Comment:
so.. is this to support both d14 and d15? or to make sure .Next.dll is not loaded when remote host is disabled?

in preview 5, remote host will be on by default, so IRemoteXXX probably won't be needed. I am going to remove IRemoteHostDiagnosticAnalyzerExecutor after preivew 5 as well. I needed it since in preview 4, the option is off by default and I had to make sure .Next.dll won't be loaded if option is not explicitly turned on.